demos/qml_optimization.ipynb

```diff
@@ -0,0 +1,469 @@
+      },
+      "outputs": [],
+      "source": [
+        "@qml.qnode(dev)\n",
+        "def circuit(data, weights):\n",
+        "    \"\"\"Quantum circuit ansatz\"\"\"\n",
+        "\n",
+        "    @catalyst.for_loop(0, n_wires, 1)\n",
+        "    def data_embedding(i):\n",
+        "        qml.RY(data[i], wires=i)\n",
+        "\n",
+        "    data_embedding()\n",
+        "\n",
+        "    @catalyst.for_loop(0, n_wires, 1)\n",
+        "    def ansatz(i):\n",
+        "        qml.RX(weights[i, 0], wires=i)\n",
+        "        qml.RY(weights[i, 1], wires=i)\n",
+        "        qml.RX(weights[i, 2], wires=i)\n",
+        "        qml.CNOT(wires=[i, (i + 1) % n_wires])\n",
+        "\n",
+        "    ansatz()\n",
+        "\n",
+        "    # we use a sum of local Z's as an observable since a\n",
+        "    # local Z would only be affected by params on that qubit.\n",
+        "    return qml.expval(qml.sum(*[qml.PauliZ(i) for i in range(n_wires)]))\n",
+        "\n",
+        "circuit = qml.qjit(catalyst.vmap(circuit, in_axes=(1, None)), autograph=False)"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "K1bpU4Au6xUE"
+      },
+      "source": [
+        "We will define a simple cost function that computes the overlap between model output and target data:"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 6,
+      "metadata": {
+        "id": "0c7bwv3IsUxP"
+      },
+      "outputs": [],
+      "source": [
+        "def my_model(data, weights, bias):\n",
+        "    return circuit(data, weights) + bias\n",
+        "\n",
+        "@qml.qjit\n",
+        "def loss_fn(params, data, targets):\n",
+        "    predictions = my_model(data, params[\"weights\"], params[\"bias\"])\n",
+        "    loss = jnp.sum((targets - predictions) ** 2 / len(data))\n",
+        "    return loss"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "qUBCtuId62o9"
+      },
+      "source": [
+        "Note that the model above is just an example for demonstration -- there are important considerations that must be taken into account when performing QML research, including methods for data embedding, circuit architecture, and cost function, in order to build models that may have use. This is still an active area of research; see our [demonstrations](https://pennylane.ai/qml/demonstrations) for details.\n",
+        "\n",
+        "## Initialize your parameters\n",
+        "\n",
+        "Now, we can generate our trainable parameters `weights` and `bias` that will be used to train our QML model."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 7,
+      "metadata": {
+        "id": "sOWgLW5esIsO"
+      },
+      "outputs": [],
+      "source": [
+        "weights = jnp.ones([n_wires, 3])\n",
+        "bias = jnp.array(0.)\n",
+        "params = {\"weights\": weights, \"bias\": bias}"
```

Comment:
Truly putting everything together 😆

Reply:
yep, I needed the pytree support for this!

Reply:
really nice to see all these features in one demo :star_struck: